core/ecschema-editing/src/Differencing/SchemaDifferenceValidator.ts

```diff
@@ -0,0 +1,502 @@
+  }
+
+  /**
+   * Visitor implementation for handling RelationshipClassDifference.
+   * @internal
+   */
+  public async visitRelationshipClassDifference(entry: RelationshipClassDifference) {
+    await this.visitClassDifference(entry);
+  }
+
+  /**
+   * Visitor implementation for handling RelationshipConstraintDifference.
+   * @internal
+   */
+  public async visitRelationshipConstraintDifference(_entry: RelationshipConstraintDifference) {
+  }
+
+  /**
+   * Visitor implementation for handling RelationshipConstraintClassDifference.
+   * @internal
+   */
+  public async visitRelationshipConstraintClassDifference(_entry: RelationshipConstraintClassDifference) {
+  }
+
+  /**
+   * Visitor implementation for handling StructClassDifference.
+   * @internal
+   */
+  public async visitStructClassDifference(entry: StructClassDifference) {
+    await this.visitClassDifference(entry);
+  }
+
+  /**
+   * Visitor implementation for handling UnitDifference.
+   * @internal
+   */
+  public async visitUnitDifference(entry: UnitDifference) {
+    await this.visitSchemaItemDifference(entry, await this._targetSchema.getItem(entry.itemName));
+  }
+
+  /**
+   * Visitor implementation for handling UnitSystemDifference.
+   * @internal
+   */
+  public async visitUnitSystemDifference(entry: UnitSystemDifference) {
+    await this.visitSchemaItemDifference(entry, await this._targetSchema.getItem(entry.itemName));
+  }
+}
+
+/**
+ * Helper method to resolve the schema item name from lazy loaded schema items.
+ * @param lazyItem  LazyLoaded item
+ * @returns         The full name of the item or undefined item was not set.
+ */
+function resolveLazyItemName(lazyItem?: LazyLoadedSchemaItem<SchemaItem>) {
+  return lazyItem && lazyItem.fullName;
+}
+
+/**
+ * Helper method to resolve the type name of a property.
+ * @param property  The property which's type shall be resolved.
+ * @returns         The (full) name of the properties type.
+ */
+function resolvePropertyTypeName(property: Property) {
+  const [prefix, suffix] = property.isArray() ? ["[", "]"] : ["", ""];
+  if (property.isEnumeration())
+    return `${prefix}${resolveLazyItemName(property.enumeration)}${suffix}`;
+  if (property.isPrimitive())
+    return `${prefix}${primitiveTypeToString(property.primitiveType)}${suffix}`;
+  if (property.isStruct())
+    return `${prefix}${property.structClass.fullName}${suffix}`;
+  if (property.isNavigation())
+    return `${prefix}${property.relationshipClass.fullName}${suffix}`;
+  return propertyTypeToString(property.propertyType);
+}
+
+/**
+ * Recursive synchronous function to figure whether a given class derived from
+ * a class with the given baseClassName.
+ */
+async function derivedFrom(ecClass: ECClass | undefined, baseClassName: string): Promise<boolean> {
+  if (ecClass === undefined) {
+    return false;
+  }
+  if (ecClass && ecClass.name === baseClassName) {
+    return true;
+  }
+  return derivedFrom(await ecClass.baseClass, baseClassName);
```

Comment:
This concerns me because classes can have the same name but come from different classes ... I know there is some wiggly bits handling schemas with different names here but we have to ensure that we will not match if base classes are from a different schema but the input class name comes from this schema.


There should be at least one test for this explicit case